balancer/base/balancer.go

```diff
@@ -90,15 +94,23 @@ func (b *baseBalancer) UpdateResolverState(s resolver.State) {
 			sc.Connect()
 		}
 	}
-	for a, sc := range b.subConns {
-		// a was removed by resolver.
+
+	// Avoid triggering regeneratePicker every time by HandleSubConnStateChange when call b.cc.RemoveSubConn
+	// Remove all conns in balancer firstly
+	// Balancer should avoid using these removed conns immediately
+	deletedConns := make([]balancer.SubConn, 0)
+	for a := range b.subConns {
 		if _, ok := addrsSet[a]; !ok {
-			b.cc.RemoveSubConn(sc)
+			deletedConns = append(deletedConns, b.subConns[a])
 			delete(b.subConns, a)
-			// Keep the state of this sc in b.scStates until sc's state becomes Shutdown.
-			// The entry will be deleted in HandleSubConnStateChange.
 		}
 	}
+
+	for _, sc := range deletedConns {
+		b.cc.RemoveSubConn(sc)
```

Comment:
Moving this code doesn't necessarily change the number of regeneratePicker() being called, because each one will still trigger a subconn state change, and the updates come asynchronously.